### PR TITLE
 MSTeams and Slack are Failsafe, implementation of Strategy Pattern for Exception Handling and other refactring

### PR DESCRIPTION
**Pull request by mithunbharadwaj**
_**Implementation Details**_
**APIExceptionStrategy:** This strategy is designed for handling exceptions that occur in the API context. It enhances API robustness by providing custom exception handling, such as logging specific errors or re-raising exceptions with API-friendly messages.

**EmailExceptionStrategy:** Focuses on handling exceptions by notifying stakeholders via email. This is particularly useful for asynchronous error reporting and alerting admin or support teams about critical issues.

Both strategies implement the handle_exception method from the ExceptionStrategy abstract base class, ensuring a consistent interface while allowing different handling mechanisms.

- MSTeams and Slack as Failsafe: We have enhanced the reliability of our notification services by implementing failsafe mechanisms for both MSTeams and Slack integrations. 

- Improved SMTP Error Handling: The SMTP service has been updated to provide more detailed error reporting. 

- Jinja2TemplateSyntaxError. This exception is specifically designed to address and report syntax errors in Jinja2 templates. A notable feature of this exception is its ability to report not only the nature of the syntax error but also the specific path to the problematic template

### CODE REVIEW

File: asabiris/orchestration/sendslack.py [hunk 1]
```diff
@@ -44,58 +47,50 @@ async def send_to_slack(self, msg):
...
-				fallback_message = self.MarkdownFormatterService.unformat(output)
-
-				# See https://api.slack.com/reference/block-kit/blocks
-				blocks = [
-					{
-						"type": "section",
-						"text": {
-							"type": "mrkdwn",
-							"text": output
+		try:
```
Comment (eliska-n) [inline asabiris/orchestration/sendslack.py:50]: No, Mithun, no "except Exception" on my watch. You must try harder.

File: asabiris/orchestration/sendslack.py [hunk 2]
```diff
@@ -44,58 +47,50 @@ async def send_to_slack(self, msg):
...
 
-		return jinja_output
+			output = self.MarkdownFormatterService.unformat(output)
+			atts_gen = self.AttachmentRenderingService.render_attachment(attachments)
+			await self.SlackOutputService.send_files(output, atts_gen)
+		except Exception as e:
+			L.exception("Error occurred when preparing slack notification")
+			error_message = self._generate_error_message_slack(str(e))
+			blocks = None
+			await self.SlackOutputService.send_message(blocks, error_message)
```
Comment (eliska-n) [inline asabiris/orchestration/sendslack.py:93]: what if the exception comes from sending the message? then, you send another message which does not reach its target and you return 200 in response?

That's not a good approach

File: asabiris/orchestration/sendslack.py [hunk 3]
```diff
@@ -44,58 +47,50 @@ async def send_to_slack(self, msg):
...
-		"""
+				await self.SlackOutputService.send_message(blocks, fallback_message)
+				return
 
-		try:
-			jinja_output = await self.JinjaService.format(template, params)
-		except KeyError:
-			L.warning("Failed to load or render a template (missing?)", struct_data={'template': template})
-			raise
+			# Sending attachments
```
Comment (eliska-n) [inline asabiris/orchestration/sendslack.py:84]: can you send a message with attachment? or is it either mesage or attachment without message?

File: asabiris/orchestration/sendslack.py [hunk 4]
```diff
@@ -123,3 +118,17 @@ def get_content_type(self, file_extension):
 		"""
 		content_type = mimetypes.guess_type('dummy' + file_extension)[0]
 		return content_type or 'application/octet-stream'
+
+	def _generate_error_message_slack(self, specific_error: str) -> Tuple[str, str]:
+		timestamp = datetime.datetime.now(tz=datetime.timezone.utc).strftime("%Y-%m-%d %H:%M:%S")
+		error_message = (
+			":warning: *Hello!*\n\n"
+			"We encountered an issue while processing your request:\n`{}`\n\n"
```
Comment (eliska-n) [inline asabiris/orchestration/sendslack.py:126]: You want to send the specific error to the end user (customer!) only if they can do something about that. If it says there is wrong configuration if the BE service (for example), it is not helpful.

File: asabiris/orchestration/sendslack.py (hunk 1 above)
Comment (eliska-n) [reply asabiris/orchestration/sendslack.py:50]: You must sort the exceptions.

**Review by eliska-n — CHANGES_REQUESTED** [review]

File: asabiris/orchestration/sendslack.py [hunk 5]
```diff
@@ -44,58 +47,50 @@ async def send_to_slack(self, msg):
...
-						"text": {
-							"type": "mrkdwn",
-							"text": output
+		try:
+			# templates must be stores in /Templates/Slack
+			if not template.startswith("/Templates/Slack/"):
+				raise PathError(use_case='Slack', invalid_path=template)
+
+			params = body.get("params", {})
+			output = await self.JinjaService.format(template, params)
```
Comment (eliska-n) [inline asabiris/orchestration/sendslack.py:56]: this is string

File: asabiris/orchestration/sendslack.py [hunk 6]
```diff
@@ -44,58 +47,50 @@ async def send_to_slack(self, msg):
...
-		# Sending attachments
-
-		output = self.MarkdownFormatterService.unformat(output)
-		atts_gen = self.AttachmentRenderingService.render_attachment(attachments)
-		await self.SlackOutputService.send_files(output, atts_gen)
+					]
 
+				else:
+					# This is a plain text Slack message
+					fallback_message = output
```
Comment (eliska-n) [inline asabiris/orchestration/sendslack.py:78]: this is string

File: asabiris/orchestration/sendslack.py [hunk 7]
```diff
@@ -44,58 +47,50 @@ async def send_to_slack(self, msg):
...
+					blocks = None
 
-	async def render_attachment(self, template, params):
-		"""
-		This method renders attachment based on the depending on the extension of template.
-		"""
+				await self.SlackOutputService.send_message(blocks, fallback_message)
```
Comment (eliska-n) [inline asabiris/orchestration/sendslack.py:81]: here the second argument is string

File: asabiris/orchestration/sendslack.py [hunk 8]
```diff
@@ -123,3 +118,17 @@ def get_content_type(self, file_extension):
 		"""
 		content_type = mimetypes.guess_type('dummy' + file_extension)[0]
 		return content_type or 'application/octet-stream'
+
+	def _generate_error_message_slack(self, specific_error: str) -> Tuple[str, str]:
```
Comment (eliska-n) [inline asabiris/orchestration/sendslack.py:122]: this does not return tuple, am I right? You copypasted this from some copilot or chatgpt without reviewing the code

**Review by eliska-n — CHANGES_REQUESTED** [review]

**Comment by ateska** [comment]
No, this is nonsense ... "failsafe", "exception_manager" ... I completely don't get this.

**Comment by ateska** [comment]
This MR will be closed - after discussion - and another mechanism for error handling and forward fail-safe will be introduced.

**Pull request closed without merging**